src/java.base/share/classes/javax/crypto/KDFSpi.java

```diff
@@ -0,0 +1,147 @@
+     * that support initialization parameters.
+     *
+     * @param kdfParameters
+     *     the initialization parameters for the {@code KDF} algorithm (may be
+     *     {@code null})
+     *
+     * @throws InvalidAlgorithmParameterException
+     *     if the initialization parameters are inappropriate for this
+     *     {@code KDFSpi}
+     */
+    protected KDFSpi(KDFParameters kdfParameters)
+        throws InvalidAlgorithmParameterException {}
+
+    /**
+     * Returns the {@code KDFParameters} used with this {@code KDF} object.
+     * <p>
+     * The returned parameters may be the same that were used to initialize
+     * this {@code KDF} object, or may contain additional default or
+     * random parameter values used by the underlying KDF algorithm.
+     * If the required parameters were not supplied and can be generated by
+     * the {@code KDF} object, the generated parameters are returned;
+     * otherwise {@code null} is returned.
+     *
+     * @return the parameters used with this {@code KDF} object, or
+     * {@code null}
+     *
+     * @throws UnsupportedOperationException if this method is not overridden
+     * by a provider
+     */
+    protected abstract KDFParameters engineGetParameters();
+
+    /**
+     * Derives a key, returned as a {@code SecretKey}.
+     * <p>
+     * The {@code engineDeriveKey} method may be called multiple times on a particular
+     * {@code KDFSpi} instance.
+     *
+     * @param alg
+     *     the algorithm of the resultant {@code SecretKey} object
+     * @param derivationParameterSpec
+     *     derivation parameters
+     *
+     * @return a {@code SecretKey} object corresponding to a key built from the
+     *     KDF output and according to the derivation parameters.
+     *     Implementations must not return a {@code null} value here. Instead,
+     *     they must throw an appropriate {@code Exception} if a value cannot
+     *     be returned.
+     *
+     * @throws InvalidAlgorithmParameterException
+     *     if the information contained within the {@code derivationParameterSpec} is
+     *     invalid, if {@code alg} is invalid, or if their combination
+     *     results in something invalid, ie - a key of inappropriate length
+     *     for the specified algorithm
+     * @throws NullPointerException
+     *     if {@code alg} or {@code derivationParameterSpec} is null
+     */
+    protected abstract SecretKey engineDeriveKey(String alg,
+                                                 AlgorithmParameterSpec derivationParameterSpec)
+        throws InvalidAlgorithmParameterException;
+
+    /**
+     * Obtains raw data from a key derivation function.
+     * <p>
+     * The {@code engineDeriveData} method may be called multiple times on a
+     * particular {@code KDFSpi} instance.
+     *
+     * @param derivationParameterSpec
+     *     derivation parameters
+     *
+     * @return a byte array corresponding to a key built from the
```

Comment:
Can't we just state 

> @return a byte array output by this KDF object using the derivation parameters.

No need to mention the "key built from ..." part. KDF output is bytes, we package it into key objects and not the other way around.